tools/importer-msgraph-metadata/pipeline/parser_models.go

```diff
@@ -0,0 +1,531 @@
+					}
+					for k, v := range result.Schemas {
+						schemas[k] = v
+					}
+				}
+			}
+		}
+	}
+
+	if schema.Title != "" {
+		title = schema.Title
+	}
+
+	// prefer the innermost type
+	if typ == "" && schema.Type != "" {
+		typ = schema.Type
+	}
+
+	// prefer the innermost format
+	if format == "" && schema.Format != "" {
+		format = schema.Format
+	}
+
+	if len(schema.Enum) > 0 {
+		enum = schema.Enum
+	}
+
+	if schema.Properties != nil {
+		for k, v := range schema.Properties {
+			schemas[k] = v
+		}
+	}
+
+	if len(schemas) == 0 {
+		schemas = nil
+	}
+
+	return flattenedSchema{
+		Schemas: schemas,
+		Title:   title,
+		Type:    typ,
+		Format:  format,
+		Enum:    enum,
+	}, seenRefs
+}
+
+func parseSchemaRef(schemaRef *openapi3.SchemaRef) *openapi3.Schema {
+	if schemaRef.Value != nil {
+		return schemaRef.Value
+	}
+	return nil
+}
+
+// parseSchemas inspects the provided flattenedSchema to parse out the fields for the provided modelName, optionally
+// marking it as a common model. The provided Models (map[string]Model) is mutated to append the new model and its fields.
+func parseSchemas(input flattenedSchema, modelName string, models Models, common bool) Models {
+	if _, ok := models[modelName]; ok {
+		return models
+	}
+	model := Model{
+		Fields: make(map[string]*ModelField),
+		Common: common,
+	}
+	models[modelName] = &model
+
+	for k, v := range input.Schemas {
+		schema := parseSchemaRef(v)
+		result, _ := flattenSchema(schema, nil)
+		title := ""
+
+		if result.Title != "" {
+			title = strings.Title(result.Title)
+		} else {
+			title = strings.Title(k)
+		}
+
+		field := ModelField{
+			Title:       title,
+			Description: schema.Description,
+			Default:     schema.Default,
+			Enum:        parseEnum(schema.Enum),
+			JsonField:   k,
+		}
+
+		if len(field.Enum) == 0 && len(result.Enum) > 0 {
+			field.Enum = parseEnum(result.Enum)
+		}
+
+		if result.Schemas != nil {
+			if _, ok := models[title]; !ok {
+				models = parseSchemas(result, title, models, common)
+			}
+			field.ModelName = &title
+		}
+
+		if schema.Items != nil && schema.Items.Value != nil && schema.Items.Value.Type != "" {
+			field.ItemType = fieldType(schema.Items.Value.Type, schema.Items.Value.Format, field.ModelName != nil)
+		}
+
+		if schema.Type == "" && schema.Format == "" && (result.Type != "" || result.Format != "") {
+			field.Type = fieldType(result.Type, result.Format, field.ModelName != nil)
+		} else {
+			field.Type = fieldType(schema.Type, schema.Format, field.ModelName != nil)
+		}
+
+		if field.Type != nil && *field.Type == DataTypeArray && len(field.Enum) > 0 && (result.Type != "" || result.Format != "") {
+			field.ItemType = fieldType(result.Type, result.Format, field.ModelName != nil)
+		}
+
+		model.Fields[cleanName(k)] = &field
+	}
+
+	return models
+}
+
+// parseEnum returns a slice of sanitized enum values (which are always strings)
+func parseEnum(input []interface{}) []string {
+	out := make([]string, 0, len(input))
```

Comment:
we're appending to the slice rather than indexing it, so we'll end up with `nil` items at the end by pre-sizing this, should `raw` be nil below?

```suggestion
	out := make([]string, 0)
```